### DISCUSSION TITLE
regeneratorRuntime before v7.18.0

### DISCUSSION THREAD

**Discussion by Lancernix** [babel/babel#15483] [Q&A] [answered]
Hello, I have a strange problem about `regenerator-runtime`. I read the release document and found that after v7.18.0, when `@babel/preset-env` set `useBuiltIns: 'entry'`, there is no need to import `regenerator-runtime` in the entry file. And before v7.18.0, if the `regenerator-runtime` is not imported, a “the regeneratorRuntime is not defined” error will appear. I verified this change on [Babel's playground](https://babeljs.io/repl) and my machine.
In the playground, the compiled result was as expected：
![image](https://user-images.githubusercontent.com/26547236/224289071-be339909-dcd3-427a-9b6e-7f88e76df07e.png)
But on my machine, I did not import `regenerator-runtime` in the entry file, but there was a `_ regeneratorRuntime` function. This code can be executed successfully.
![image](https://user-images.githubusercontent.com/26547236/224290857-021544d4-6de0-4e00-868a-9744b7a285b0.png)
Why? I'm a little confused...

My env:
- MacBook Pro m1
- MacOS 13.2
- node v16.14.0

Demo git: https://github.com/Lancernix/babel-regenerator-demo

**Comment by nicolo-ribaudo** [comment, marked as answer]
`regeneratorRuntime` is now a "normal Babel helper". It's behavior is similar to, for example, the `classCallCheck` helper that Babel injects when compiling `class A {}`:

```js
function _createClass(Constructor, protoProps, staticProps) { /* ... */ }
function _classCallCheck(instance, Constructor) { /* ... */ }
var A = /*#__PURE__*/_createClass(function A() {
  _classCallCheck(this, A);
});
```

By default, Babel injects all the helpers inline to make sure that they are available.

<details>
<summary>EDIT: hiding this part of the answer because I misread your question, it's a bit off topic</summary>

If you have multiple files and you want to avoid having the same helpers duplicated in every single file where they are used, you can use `@babel/plugin-transform-runtime`:
1. Install `@babel/runtime` in your dependencies, which is a package that contains all the runtime Babel helpers
2. Install `@babel/plugin-transform-runtime` to your dev dependencies (which is where you should usualy install all the `@babel/*` packages, since you only need them at dev time — except for `@babel/runtime`)
3. Add this plugin to your Babel config:
   ```json5
   "plugins": [
     ["@babel/plugin-transform-runtime", { "version": "7.21.0" }]
   ]
   ```
   where instead of 7.21.0 you write the version you installed of the `@babel/runtime` package.

</details>

The reason you are seeing the new behavior even with the old `@babel/core` and `@babel/preset-env` set to 7.17 is that the automatic injection of `@babel/runtime` has been implemented in _`@babel/plugin-transform-regenerator`_ 7.18.0: if you check in your `package-lock.json` file, you have a modern version of that package.

**Reply by Lancernix** [reply]
Thanks! I checked`package-lock.json` and found that the version of `@babel/plugin-transform-regenerator` is v7.20.5. I manually locked  the version to v7.17.9 and got the same compiled result as Babel playground. So actually, v7.18.0 means the version of `@babel/plugin-transform-regenerator`!
